### PR TITLE
openssl_3_1: init at 3.1.3

### PR DESCRIPTION
**Pull request by thillux**
OpenSSL 3.1 is the most recent release to develop against, while OpenSSL 3.0 is a LTS release, most developers should probably choose now (see:
https://github.com/openssl/openssl/issues/20722).

Add OpenSSL 3.1.3 in order to allow development against this version with Nix. (I started with 3.1.1 here, but then came the next security update).

Currently OpenSSL 3.0 and 3.1 are independent release lines. In my opinion, some OpenSSL changes between release lines are rather subtle and it should be up to the developer, when to upgrade the main release line at his discretion. So openssl_3 should not just be set to have version 3.1.1 in my opinion.

My current use case is to develop a random number generator provider and test it against both OpenSSL 3.0 and 3.1.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thillux** [comment]
Result of `nixpkgs-review pr 244893` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>openssl_3_1</li>
    <li>openssl_3_1.bin</li>
    <li>openssl_3_1.debug</li>
    <li>openssl_3_1.dev</li>
    <li>openssl_3_1.doc</li>
    <li>openssl_3_1.man</li>
  </ul>
</details>

File: pkgs/development/libraries/openssl/default.nix [hunk 1]
```diff
@@ -262,4 +262,26 @@ in {
       license = licenses.asl20;
     };
   };
+
+  openssl_3_1 = common {
+    version = "3.1.1";
+    sha256 = "sha256-s6phM0IzuFK2PdsEjfGBF3wsZZ651DdgCBGPnAjQdnQ=";
```
Comment (RaitoBezarius) [inline pkgs/development/libraries/openssl/default.nix:268]: Please let's use `hash` if possible

**Comment by RaitoBezarius** [comment]
Is it possible to add yourself as a OpenSSL maintainer?

**Comment by thillux** [comment]
> Is it possible to add yourself as a OpenSSL maintainer?

Sure, let's pray my first CVE does not fly in by tomorrow :)

File: pkgs/development/libraries/openssl/default.nix (hunk 1 above)
Comment (thillux) [reply pkgs/development/libraries/openssl/default.nix]: I switched to hash.

**Comment by thillux** [comment]
@RaitoBezarius I updated to the current security fix release.

**Comment by RaitoBezarius** [comment]
Will test and merge if everything is okay shortly.

**Comment by thillux** [comment]
We should probably wait for 1.1.1v and 3.0.10 to be merged. Afterwards I can change the hashes here accordingly.

**Comment by vcunat** [comment]
It will take at least a week for those to hit nixpkgs master, probably.

**Comment by vcunat** [comment]
This was also discussed on PR #221206

For me the main point is that 3.x [are supposed to be](https://www.openssl.org/policies/releasestrat.html) "API/ABI compatible feature releases".  In that case I'm for switching default to 3.1 on nixpkgs unstable/master.  We could keep a non-default 3.0 as well, that's rather cheap and perhaps someone will find it useful.

**Comment by vcunat** [comment]
(I don't really care about that bit.  For my direct development it's gnutls everywhere instead.)

**Comment by ajs124** [comment]
> This was also discussed on PR #221206

You've already linked it, but to quote myself "I intend to keep OpenSSL on LTS releases".

Obviously, I'm not even listed as a maintainer, so I don't really have any say in this and I won't try to block you, but that's what I suggest anyone to do that feels like maintaining openssl in nixpkgs.

When I tried bumping to 3.1.x in #221206, I couldn't even manage to build up to systemd.

**Review by SuperSandro2000 — COMMENTED** [review]
This PR rebuilds a lot of packages which means we must target staging. Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging) to not potentially ping a lot of people.

**Comment by vcunat** [comment]
No, it does not rebuild a lot of packages.  That was just a detection artifact, because you touched the default openssl.

**Comment by luzpaz** [comment]
3.1.3 has been released https://github.com/openssl/openssl/releases/tag/openssl-3.1.3

**Comment by thillux** [comment]
> 3.1.3 has been released https://github.com/openssl/openssl/releases/tag/openssl-3.1.3

I updated the PR accordingly.

File: pkgs/development/libraries/openssl/default.nix [hunk 2]
```diff
@@ -256,7 +257,29 @@ in {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/openssl/default.nix:257]: the sha265 of openssl_1_1 needs to be changed to hash, too

Comment (thillux) [reply pkgs/development/libraries/openssl/default.nix:257]: Thanks for the hint. It got lost during a rebase.

**Comment by RaitoBezarius** [comment]
> > We could keep a non-default 3.0 as well, that's rather cheap and perhaps someone will find it useful.
> 
> I would rather not unless somebody specifically asks for it and has a good reason for it.

> > This was also discussed on PR #221206
> 
> You've already linked it, but to quote myself "I intend to keep OpenSSL on LTS releases".
> 
> Obviously, I'm not even listed as a maintainer, so I don't really have any say in this and I won't try to block you, but that's what I suggest anyone to do that feels like maintaining openssl in nixpkgs.
> 
> When I tried bumping to 3.1.x in #221206, I couldn't even manage to build up to systemd.

@alyssais Going back to what was discussed here, are you okay with this reason for having a non-default 3.1 ?

I believe we ought to retry in the future to bump to default 3.1 and see at what time do we switch to a strategy where we pin stuff to 3.0.

But IMHO, this should not block this PR any further, I am inclined to merge and move the discussion on defaulting to 3.1 somewhere else while taking into account @ajs124 position.

We have the chance to have multiple OpenSSL maintainers now, and I feel like we can do interesting things regarding OpenSSL maintenance, do we have any rush figuring out how to do so?

**Comment by alyssais** [comment]
> @alyssais Going back to what was discussed here, are you okay with this reason for having a non-default 3.1 ?

Yeah, that's fine.  It means that the OpenSSL maintainers have broken their compatibility promise for 3.x, but there's nothing we can do about that.

**Pull request merged**